### PR TITLE
tests.vim: init (moved from vim-utils.nix)

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
Remove hardcoded defaults (especially as it may be problematic):
https://github.com/NixOS/nixpkgs/issues/118953
I would also like to make it possible to avoid wrriting the init.vim / vimrc file when the config is empty see https://github.com/nix-community/home-manager/issues/1907

Ideally I would like to write more tests for the vimrc generators but this would require a home-manager -like infrastructure with functions like `assertFileRegex` to detect that the generated vimrc contains a path to a folder with vimplugins for instance.

###### Things done
Moved tests from pkgs/misc/vim-plugins/vim-utils.nix to pkgs/test/vim.
Also reduced the amount of generated config.

- Make it possible to have an empty config when configured adequately
- removed default vim config when using native packages, it could be
  source of bugs see linked issues (syntax on overrides vim highlights)

you can run the tests via
$ nix-build -A tests.vim

cc @timokau @doronbehar @mjlbach 

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119467` at 1be8e91a run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>gnvim</li>
    <li>neovim-qt</li>
    <li>nvimpager</li>
    <li>spacevim</li>
  </ul>
</details>

File: pkgs/misc/vim-plugins/vim-utils.nix [hunk 1]
```diff
@@ -183,13 +184,51 @@ let
...
+      ++ ["mkdir -p $out/pack/${packageName}/opt"]
+      ++ (builtins.map (link packageName "opt") opt)
+    );
+    packDir = (packages:
+      stdenv.mkDerivation {
+        name = "vim-pack-dir";
+        src = ./.;
+        installPhase = lib.concatStringsSep
+                          "\n"
+                          (lib.flatten (lib.mapAttrsToList packageLinks packages));
```
Comment (SuperSandro2000) [inline pkgs/misc/vim-plugins/vim-utils.nix:212]:
```suggestion
        installPhase = lib.concatStringsSep "\n" (lib.flatten (lib.mapAttrsToList packageLinks packages));
```

File: pkgs/misc/vim-plugins/vim-utils.nix [hunk 2]
```diff
@@ -183,13 +184,51 @@ let
...
+    packageLinks = (packageName: {start ? [], opt ? []}:
+    let
+      # `nativeImpl` expects packages to be derivations, not strings (as
+      # opposed to older implementations that have to maintain backwards
+      # compatibility). Therefore we don't need to deal with "knownPlugins"
+      # and can simply pass `null`.
+      depsOfOptionalPlugins = lib.subtractLists opt (findDependenciesRecursively opt);
+      startWithDeps = findDependenciesRecursively start;
+    in
+      ["mkdir -p $out/pack/${packageName}/start"]
```
Comment (SuperSandro2000) [inline pkgs/misc/vim-plugins/vim-utils.nix:199]:
```suggestion
      [ "mkdir -p $out/pack/${packageName}/start" ]
```

File: pkgs/misc/vim-plugins/vim-utils.nix [hunk 3]
```diff
@@ -183,13 +184,51 @@ let
...
+      # To avoid confusion, even dependencies of optional plugins are added
+      # to `start` (except if they are explicitly listed as optional plugins).
+      ++ (builtins.map (link packageName "start") (lib.unique (startWithDeps ++ depsOfOptionalPlugins)))
+      ++ ["mkdir -p $out/pack/${packageName}/opt"]
```
Comment (SuperSandro2000) [inline pkgs/misc/vim-plugins/vim-utils.nix:203]:
```suggestion
      ++ [ "mkdir -p $out/pack/${packageName}/opt" ]
```

File: pkgs/test/vim/configs.nix [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# nix-instantiate --eval --strict . -A tests.vimConfigs
+{ lib, pkgs }:
```
Comment (SuperSandro2000) [inline pkgs/test/vim/configs.nix:2]: pkgs is unused, right?

File: pkgs/test/vim/default.nix [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#
+#
```
Comment (SuperSandro2000) [inline pkgs/test/vim/default.nix:2]:
```suggestion
```

File: pkgs/test/vim/configs.nix (hunk 4 above)
Comment (teto) [reply pkgs/test/vim/configs.nix]: this file shoundlnt have been committed, it's not used, I've just removed it. thanks

**Review by mjlbach — APPROVED** [review]
LGTM, tested and the changes appear are minor/improve organization.

**Comment by teto** [comment]
Things to watch out for:
- if you set configure.beforePlugins yourself, you will need to add `set nocompatible` too not to lose it
- `filetype indent plugin on | syn on` is not enabled by default for the vim-plug installer: one should not override the default of vim, at least not make it mandatory. Also sometimes it's enabled before plugins (pathogen etc,) which is not consistent

**Pull request merged**